### PR TITLE
Fix upsample_bilinear2d decomp for uint8

### LINKED ISSUE #113024: Upsample bilinear 2d decomposition does not match native implementation for uint8
### 🐛 Describe the bug

The `torch.nn.functional.interpolate` op in its implementation will call `torch._C.nn.upsample_bilinear2d` when `mode=bilinear` and `input.dim() == 4`. However when the decomposition for `upsample_bilinear2d` is called, as opposed to the native implementation of `torch._C.nn.upsample_bilinear2d`, the results of `interpolate` differ.

After doing some digging, I can see the difference comes from the results of `torch._C.nn.upsample_bilinear2d` and its respective decomposition `torch._decomps.decompositions.upsample_bilinear2d_vec` not matching. Reproducer below
```python
import torch
from torch._decomp.decompositions import upsample_bilinear2d_vec

torch.manual_seed(42)

sample_input = torch.randint(0, 255, size=[2, 3, 270, 270], 
                            dtype=torch.uint8)
size = (130, 130)

native_result = torch._C._nn.upsample_bilinear2d(sample_input, 
                                                 size, False, None)
decomp_result = upsample_bilinear2d_vec(sample_input, 
                                        size, False, None)

assert torch.eq(native_result, decomp_result).all() # fails

```

This happens with the input is of dtype `uint8`. The expected behaviour I would like to see is that the decomposition produces the same output as the native implementation.

### Versions

```
PyTorch version: 2.1.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 16.0.6 (https://github.com/conda-forge/clangdev-feedstock 60e0ed07baadaf59c7646015aff34fc0352327f0)
CMake version: version 3.25.3
Libc version: glibc-2.35

Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-57-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          256
On-line CPU(s) list:             0-255
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7742 64-Core Processor
CPU family:                      23
Model:                           49
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       2
Stepping:                        0
BogoMIPS:                        4491.60
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
Virtualization:                  AMD-V
L1d cache:                       4 MiB (128 instances)
L1i cache:                       4 MiB (128 instances)
L2 cache:                        64 MiB (128 instances)
L3 cache:                        512 MiB (32 instances)
NUMA node(s):                    8
NUMA node0 CPU(s):               0-15,128-143
NUMA node1 CPU(s):               16-31,144-159
NUMA node2 CPU(s):               32-47,160-175
NUMA node3 CPU(s):               48-63,176-191
NUMA node4 CPU(s):               64-79,192-207
NUMA node5 CPU(s):               80-95,208-223
NUMA node6 CPU(s):               96-111,224-239
NUMA node7 CPU(s):               112-127,240-255
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] torch==2.1.0+cpu
[pip3] torchvision==0.16.0+cpu
[conda] numpy                     1.26.0          py310hb13e2d6_0    conda-forge
[conda] torch                     2.1.0+cpu                pypi_0    pypi
[conda] torchvision               0.16.0+cpu               pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by isuruf**
Fixes #113024

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116770
* #116832

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -3355,20 +3355,30 @@ def upsample_bilinear2d(
...
         v1 = v1.to(dtype)
         v2 = v2.to(dtype)
         v3 = v3.to(dtype)
         v4 = v4.to(dtype)
 
-    yscale = (y_f32 - y).clamp(0.0, 1.0).to(dtype)
-    xscale = (x_f32 - x).clamp(0.0, 1.0).to(dtype)
+    def mul(b, a):
+        # Weights computation for uint8_t and multiplication trick
+        # Ported _compute_indices_int16_weights_aa from aten/src/ATen/native/cpu/UpSampleKernel.cpp
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:3370]: Nit, leave a permalink to the relevant comment in the code so that it's easier to find in the future.

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -3355,20 +3355,30 @@ def upsample_bilinear2d(
     v3 = aten._unsafe_index(input, [None, None, yp1, x])
     v4 = aten._unsafe_index(input, [None, None, yp1, xp1])
 
-    dtype = torch.float32 if not input.is_floating_point() else input.dtype
+    yscale = (y_f32 - y).clamp(0.0, 1.0).to(dtype)
+    xscale = (x_f32 - x).clamp(0.0, 1.0).to(dtype)
+
+    dtype = torch.int64 if not input.is_floating_point() else input.dtype
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:3361]: `int64`? Given that the trick actually works with `int16`s (but we don't support `int16`s I don't think) it should be enough to use `int32`, which in turn we know how to vectorise on C++. We don't have vectorisation for double size types.

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -4617,6 +4617,17 @@ def fn(a):
 
         self.common(fn, (torch.randn([3, 3, 6, 12]),))
 
+    def test_upsample_bilinear2d_uint8(self):
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:4620]: NO need to add an inductor test. Decompositions are tested through opinfos, so just add a new sample to the relevant OpInfo.

**Review by lezcano — COMMENTED** [review]
Slick! Can you post the diff of the generated code, just to make sure everything fuses as expected and all that? I think the current approach disables vectorisation (although I'm not sure if we have vectorisation for `uint8_t`, so yeah...

**Review by lezcano — COMMENTED** [review]
Btw, do we now need the `result.round()` at the end of this function?

File: test/inductor/test_torchinductor.py (hunk 3 above)
Comment (isuruf) [reply test/inductor/test_torchinductor.py:4620]: There is a test case in https://github.com/pytorch/pytorch/blob/faea6f2c7a4d6e6e7ebe21752eaa7b0ad05afa79/torch/testing/_internal/common_methods_invocations.py#L4596, but in `reference_inputs_func` instead of `sample_inputs_func`. Not sure if I should move that to `sample_inputs_func` or make the tests use `reference_inputs` instead of `sample_inputs`.

**Comment by isuruf** [comment]
> Btw, do we now need the result.round() at the end of this function?

No, because `@pw_cast_for_opmath` does it automatically.

File: test/inductor/test_torchinductor.py (hunk 3 above)
Comment (lezcano) [reply test/inductor/test_torchinductor.py:4620]: We can:
1. Move that test case to sample inputs in this PR
2. Run the decomposition tests (the forward, not the backward) on both sample inputs and reference inputs. Let's do this in a different PR as we are surely going to uncover bugs when doing so.

**Comment by lezcano** [comment]
> No, because @pw_cast_for_opmath does it automatically.

My point is, can we remove the rounding at the end of this function?

**Review by lezcano — COMMENTED** [review]
Your current errors may be coming from the fact that `>>` may not support dynamic shapes? (I haven't checked). If this is the case, this should be fixable by adding a metafunction for the relevant ops.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -4576,9 +4576,6 @@ def shape(size, rank, with_batch_channel=True):
         yield SampleInput(make_arg(shape(D, rank)), scale_factor=1.7)
         yield SampleInput(make_arg(shape(D, rank)), scale_factor=0.6)
 
-def reference_inputs_upsample(mode, self, device, dtype, requires_grad, **kwargs):
```
Comment (lezcano) [inline torch/testing/_internal/common_methods_invocations.py:4579]: The sizes of the inputs are too large. You'd need to make them smaller / keep the reference input as-is and add a small sample input that tests this path.

File: test/test_decomp.py [hunk 5]
```diff
@@ -781,7 +787,14 @@ def do_cross_ref(self, device, dtype, op, *, run_all):
                 kwargs = sample_input.kwargs
                 with self.DecompCrossRefMode(self, self.precision, self.rel_tol, dtype, run_all)\
                      as mode, enable_python_dispatcher():
-                    func(*args, **kwargs)
+                    decomposed = func(*args, **kwargs)
+
+                if aten_name in py_impl_decomposition_names:
```
Comment (lezcano) [inline test/test_decomp.py:792]: This is way too special-cased. We should make this generic to any function that is registered through `py_impl`, rather than manually keeping a list of these functions.

Chances are we should also test properly their backwards.

This change is non-trivial, so it should go in its own PR. You can just prepend a PR to this one.

**Comment by vfdev-5** [comment]
@isuruf pytorch nightly is already working on uint8 input. Repro code from the issue gives max abs error = 1 which is ok due roundings errors for uint8. We have a similar error for aten op when compare: 1) pure uint8 implementation vs 2) uint8 input -> cast to float -> apply interpolation -> round -> cast back to uint8.

Fixed point multiplication that you are applying in this PR should be used for CPU only with the only purpose to vectorize compiled code == optimization trick.

**Comment by isuruf** [comment]
Got it. I'll work on https://github.com/pytorch/pytorch/pull/116832 first. This PR was not about performance anyway.

**Comment by lezcano** [comment]
Why is this one closed? This PR is still highly relevant

**Comment by lezcano** [comment]
I don't think that a rounding error for `uint8` of 1 is "ok". It's fine if we can't fix it, but if we can, we should not incur these errors.

**Comment by isuruf** [comment]
In this one, I tried getting the decomposition to match the eager case. It worked for almost all the tests, but there was a off-by-one error in some tests still.
